AMM/contracts/AMM/src/main.sw

```diff
@@ -0,0 +1,62 @@
+contract;
+
+dep errors;
+dep events;
+dep utils;
+
+use errors::InitError;
+use events::{DefineValidExchangeEvent, RegisterPoolEvent};
+use libraries::{AMM, Exchange};
+use std::{constants::BASE_ASSET_ID, contract_id::ContractId, logging::log, storage::StorageMap};
+use utils::bytecode_root;
+
+storage {
+    /// The valid exchange contract bytecode root
+    exchange_bytecode_root: Option<b256> = Option::None,
+    /// Map that stores pools, i.e., asset identifier pairs as keys and corresponding exchange contract identifiers as values
+    pools: StorageMap<(ContractId, ContractId), ContractId> = StorageMap {},
+}
+
+impl AMM for Contract {
+    #[storage(read, write)]
+    fn initialize(exchange_contract_id: ContractId) {
+        require(storage.exchange_bytecode_root.is_none(), InitError::AlreadyInitialized);
+        let root = bytecode_root(exchange_contract_id);
+        storage.exchange_bytecode_root = Option::Some(root);
+        log(DefineValidExchangeEvent { root });
+    }
+
+    #[storage(read, write)]
+    fn add_pool(asset_pair: (ContractId, ContractId), pool: ContractId) {
+        require(storage.exchange_bytecode_root.is_some(), InitError::NotInitialized);
+        require(storage.exchange_bytecode_root.unwrap() == bytecode_root(pool), InitError::ExchangeContractBytecodeRootInvalid);
+        let exchange_contract = abi(Exchange, pool.into());
+        let pool_info = exchange_contract.pool_info();
+        require(pool_info.asset_a_id == asset_pair.0 && pool_info.asset_b_id == asset_pair.1, InitError::ExchangeContractDoesNotMatchPair);
```

Comment:
Would it be better to move this require below the `let ordered_asset_pair = ...` and check that `pool_info.asset_a_id == ordered_asset_pair.0` etc ?

Reply:
Good catch. The AMM contract stores the pairs in this order, while the exchange contract does not. I added a test case for adding a pool by providing a pair in a reverse order, and fixed the require condition.

Reply:
So will there be two exchange contracts for each asset pair?  Such as one contract for swapping from x->y and one contract for swapping from y->x.  I guess I'm confused why the exchange contracts don't store ordered asset pairs.

Reply:
There's only one exchange contract per pair. The AMM registry allows adding a pair only once, independent of order. So if someone tries to add an exchange A:B, and an exchange B:A already exists, it will revert.

Ordering the pairs in the exchange contract doesn't stop someone from adding both variants.

Reply:
I see that makes sense thank you!